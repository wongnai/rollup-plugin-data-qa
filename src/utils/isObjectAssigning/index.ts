export default function isObjectAssigning(node: Record<string, any>) {
	return node.callee?.object?.name === 'Object' && node.callee?.property?.name === 'assign'
}
