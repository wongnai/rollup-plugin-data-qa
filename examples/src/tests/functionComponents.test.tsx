import React, { useCallback, useMemo } from 'react'
import { render, screen } from '../utils/testUtils'

describe('function components', () => {
	it('01 - simple arrow function', () => {
		const ComponentA = () => <div>xxx</div>

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
	})

	it('02 - simple return', () => {
		const ComponentA = () => {
			return <div>xxx</div>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
	})

	it('03 - simple component with some properties', () => {
		const ComponentA = () => {
			return (
				<div className="bello" id="bello">
					xxx
				</div>
			)
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
	})

	it('04 - with a const element in function scope', () => {
		const ComponentA = () => {
			const B = <div>xxx</div>

			return <div>{B}</div>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getByDataQa('b')).toBeVisible()
	})

	it('05 - with a useCallback element in function scope', () => {
		const ComponentA = () => {
			const B = useCallback(() => <div>xxx</div>, [])

			return <div>{B()}</div>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getByDataQa('b')).toBeVisible()
	})

	it('06 - with a useMemo in function scope', () => {
		const ComponentA = () => {
			const B = useMemo(() => <div>xxx</div>, [])

			return <div>{B}</div>
		}
		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getByDataQa('b')).toBeVisible()
	})

	it('07 - with a function component child element', () => {
		const ComponentA = () => {
			const B = () => <div>xxx</div>

			return (
				<div>
					<B />
				</div>
			)
		}
		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getByDataQa('b')).toBeVisible()
	})

	it('08 - with list of elements in useMemo', () => {
		const data = ['xxx', 'yyy']
		const ComponentA = () => {
			const list = useMemo(() => data.map(item => <div key={item}>{item}</div>), [])

			return <div>{list}</div>
		}
		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
		expect(screen.getByDataQa('component-a')).toBeVisible()
		expect(screen.getAllByDataQa('list')).toHaveLength(2)
	})

	it('09 - fragment with list of elements', () => {
		const data = ['xxx', 'yyy']
		const ComponentA = () => {
			return (
				<>
					{data.map(item => (
						<div key={item}>{item}</div>
					))}
				</>
			)
		}
		const { container } = render(<ComponentA />)

		expect(container.children).toMatchSnapshot()
		expect(screen.queryAllByDataQa(/.*/g)).toHaveLength(0)
	})

	it('10 - fragment with list of elements in a const', () => {
		const data = ['xxx', 'yyy']
		const ComponentA = () => {
			const list = data.map(item => <div key={item}>{item}</div>)

			return <>{list}</>
		}
		const { container } = render(<ComponentA />)

		expect(container.children).toMatchSnapshot()
		expect(screen.queryAllByDataQa('list')).toHaveLength(2)
	})

	it('11 - fragment with list of elements in useMemo', () => {
		const data = ['xxx', 'yyy']
		const ComponentA = () => {
			const list = useMemo(() => data.map(item => <div key={item}>{item}</div>), [])

			return <>{list}</>
		}
		const { container } = render(<ComponentA />)

		expect(container.children).toMatchSnapshot()
		expect(screen.queryAllByDataQa('list')).toHaveLength(2)
	})

	it('12 - with a function component outside of function scope', () => {
		const ComponentB = ({ children }: { children: React.ReactNode }) => {
			return <div>{children}</div>
		}

		const ComponentA = () => {
			return <ComponentB>xxx</ComponentB>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('13 - with a function component with spreaded props', () => {
		const ComponentB = ({ children, ...props }: any) => {
			return <div {...props}>{children}</div>
		}

		const ComponentA = () => {
			return <ComponentB>xxx</ComponentB>
		}

		const { container } = render(<ComponentA />)

		expect(container.firstChild).toMatchSnapshot()
	})
})
