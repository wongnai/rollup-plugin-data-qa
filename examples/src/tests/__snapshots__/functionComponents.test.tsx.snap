// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`function components > 01 - simple arrow function 1`] = `
<div
  data-qa="component-a"
>
  xxx
</div>
`;

exports[`function components > 02 - simple return 1`] = `
<div
  data-qa="component-a"
>
  xxx
</div>
`;

exports[`function components > 03 - simple component with some properties 1`] = `
<div
  class="bello"
  data-qa="component-a"
  id="bello"
>
  xxx
</div>
`;

exports[`function components > 04 - with a const element in function scope 1`] = `
<div
  data-qa="component-a"
>
  <div
    data-qa="b"
  >
    xxx
  </div>
</div>
`;

exports[`function components > 05 - with a useCallback element in function scope 1`] = `
<div
  data-qa="component-a"
>
  <div>
    xxx
  </div>
</div>
`;

exports[`function components > 06 - with a useMemo in function scope 1`] = `
<div
  data-qa="component-a"
>
  <div>
    xxx
  </div>
</div>
`;

exports[`function components > 07 - with a function component child element 1`] = `
<div
  data-qa="component-a"
>
  <div
    data-qa="b"
  >
    xxx
  </div>
</div>
`;

exports[`function components > 08 - with list of elements in useMemo 1`] = `
<div
  data-qa="component-a"
>
  <div>
    xxx
  </div>
  <div>
    yyy
  </div>
</div>
`;

exports[`function components > 09 - fragment with list of elements 1`] = `
HTMLCollection [
  <div>
    xxx
  </div>,
  <div>
    yyy
  </div>,
]
`;

exports[`function components > 10 - fragment with list of elements in a const 1`] = `
HTMLCollection [
  <div>
    xxx
  </div>,
  <div>
    yyy
  </div>,
]
`;

exports[`function components > 11 - fragment with list of elements in useMemo 1`] = `
HTMLCollection [
  <div>
    xxx
  </div>,
  <div>
    yyy
  </div>,
]
`;

exports[`function components > 12 - with a function component outside of function scope 1`] = `
<div
  data-qa="component-b"
>
  xxx
</div>
`;

exports[`function components > 13 - with a function component with spreaded props 1`] = `
<div
  data-qa="component-a"
>
  xxx
</div>
`;
